### ISSUE TITLE
Why is line slower than points?

### ISSUE THREAD

**Issue by ceball** [bokeh/datashader#456]
I started with examples/timeseries.ipynb, modifying to have only one column:

<img width="808" alt="screen shot 2017-09-08 at 11 31 52 pm" src="https://user-images.githubusercontent.com/1929/30233519-ea8724b2-94ed-11e7-8135-fd5770482e66.png">

In a bit of an ugly process, I then time `canvas.line()` and `canvas.points()` for increasing repeats of the dataframe:

<img width="1007" alt="screen shot 2017-09-08 at 11 33 20 pm" src="https://user-images.githubusercontent.com/1929/30233568-370f13da-94ee-11e7-84bf-ea0de713b563.png">

line seems to be a lot slower than points:

<img width="562" alt="screen shot 2017-09-08 at 11 46 38 pm" src="https://user-images.githubusercontent.com/1929/30233818-f4827b68-94ef-11e7-9e68-141f80cb4104.png">

<img width="608" alt="screen shot 2017-09-08 at 11 47 37 pm" src="https://user-images.githubusercontent.com/1929/30233840-1a3b9b0a-94f0-11e7-9d4d-f16a699ea081.png">

(Sorry for the screenshots, but github doesn't seem to allow notebooks to be attached.)

**Comment by ceball** [comment]
I just looked at the source code and see line uses `agg=any()` by default, while points uses `agg=count()` by default. So just a note to confirm that the findings above don't change much if I use the same reduction for both line and points.

**Comment by jbednar** [comment]
Joseph or Greg had some ideas about how to speed it up, but fundamentally
lines should always be slower than points because points just increments
one bin, while line needs to solve the equation for a line and increment
every point along the way.

On Sep 8, 2017 6:09 PM, "Chris B" <notifications@github.com> wrote:

I just looked at the source code and see line uses agg=any() by default,
while points uses agg=count() by default. So just a note to confirm that
the findings above don't change much if I use the same reduction for both
line and points.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bokeh/datashader/issues/456#issuecomment-328234181>,
or mute
the thread

**Comment by ceball** [comment]
Yes, I guess the title should have been more like, 'Is line the expected amount slower than points?', or maybe, 'Can line be made faster?'.

**Comment by jbcrail** [comment]
My immediate suggestions for performance gains are:

1. Switch to integer-specific Bresenham line-drawing algorithm

We currently use the more general algorithm that supports both ints and floats. However, by the time, we need to draw a line, the points have been mapped to an integer space so float support is redundant.

2. Switch line-clipping algorithm

We currently use Cohen-Sutherland for clipping a line to a bounding box, but the Liang-Barsky algorithm is considered significantly more efficient.

**Comment by jbednar** [comment]
I'd be very happy to try both suggestions; we're using line a lot in some performance-critical applications, so anything will help...

**Comment by jbednar** [comment]
It would be nice to re-run those benchmarks now that #495 has been merged.
